test/test_dataloader.py

```diff
@@ -558,51 +624,92 @@ def _is_process_alive(pid, pname):
-        # simply raises an error. Both cases are interesting because
-        # 1. In case of it is KILL-ed by OS, the workers need to automatically
-        #    discover that their parent is dead and exit gracefully.
-        # 2. In case of it raises an error itself, the parent process needs to
-        #    take care of exiting the worker and then exits itself gracefully.
-        for raise_error in (True, False):
-            worker_pids = mp.Array('i', [0] * 4)
-
-            main_exit_event = mp.Event()
-            p = mp.Process(target=TestDataLoader._main_process,
-                           args=(self.dataset, worker_pids, main_exit_event, raise_error))
-            p.start()
-            worker_pids[-1] = p.pid
+    def test_proper_exit(self):
+        r'''There might be ConnectionResetError or leaked semaphore warning
+        (due to dirty process exit), but they are all safe to ignore'''
+
+        # TODO: test the case where the pin_memory_thread triggers an
+        #       error/fatal signal. I haven't found out how to properly do that.
 
-            main_exit_event.wait()
+        # Array to store the worker pids.
+        worker_pids = mp.Array('i', [-1 for _ in range(10)])
 
-            exit_status = [False] * len(worker_pids)
+        def wait_pids(pids, timeout):
+            r"""Wait for all process specified in pids to exit in given timeout."""
+            exit_status = [False for _ in pids]
             start_time = time.time()
             pname = 'python'
             while True:
-                for i in range(len(worker_pids)):
-                    pid = worker_pids[i]
+                for i in range(len(pids)):
+                    pid = pids[i]
                     if not exit_status[i]:
                         if not TestDataLoader._is_process_alive(pid, pname):
                             exit_status[i] = True
                 if all(exit_status):
                     break
                 else:
-                    if time.time() - start_time > MANAGER_STATUS_CHECK_INTERVAL + JOIN_TIMEOUT:
-                        self.fail('subprocess not terminated')
-                    time.sleep(1)
-            p.join(MANAGER_STATUS_CHECK_INTERVAL + JOIN_TIMEOUT - (time.time() - start_time))
-            self.assertFalse(p.is_alive(), 'main process not terminated')
+                    if time.time() - start_time > timeout:
+                        break
+                    time.sleep(0.5)
+            return exit_status
+
+        for use_workers, pin_memory, hold_iter_reference in itertools.product([True, False], repeat=3):
+            # `hold_iter_reference` specifies whether we hold a reference to the
+            # iterator. This is interesting because Python3 error traces holds a
+            # reference to the frames, which hold references to all the local
+            # variables. It is important to see that the processes still exit.
```

Comment:
"variables *, which include the error trace*." Otherwise it might not be immediately clear why is that a problem.

Reply:
The problem is that the iterator dtor is not called (not ref cycle), but yeah, I should make it clearer!